Protractor migration guide (#2812)
cypress-io/cypress-documentation: source/guides/references/protractor-migration-guide.md

---
title: Protractor Migration Guide
---

{% note info %}
# {% fa fa-graduation-cap %} What you'll learn

- Benefits of using Cypress in Angular apps
- How Cypress can create reliable e2e tests for Angular apps
- How to migrate Protractor tests into Cypress
{% endnote %}

> ## Introduction

Reviewer: Main headings should start with h1 `#`, see main note about nesting headings suggestion.

```suggestion
# Introduction
```

Author: Got it. In the future, we may want to look into restructuring how the sidebar gets generated since this leads to multiple h1's on the page which makes it difficult for screen readers and accessibility content-wise.

Considering using Cypress but unsure of whether the migration is worth it for your Angular app? We got you covered here with this migration guide to help transition from Protractor to Cypress. 

{% note warning %}

> If you see any inaccuracies with this or feel something has been misrepresented, please [submit an issue here](https://github.com/cypress-io/cypress-documentation/issues/new).

Author: Thanks for letting me know!

## Benefits of Using Cypress

As many developers can attest to, end-to-end testing is one of those things that they know they *should* do; but often do not. Or if they do run tests, the tests are often flaky and often very expensive due to how long it can take to run. And while there are often ideals of complete code coverage, the realities of business and deadlines often take precedence and the tests are left unwritten, or worse, ignored when errors are being reported because they are not reliable. With Cypress, not only will your tests be reliable, but it provides developers with tools that make e2e testing an asset to development rather than a hindrance.

### Interact with your tests in a browser

When Protractor runs tests, the browser automation launches a browser instance and often runs through tests too fast for the human eye. Without additional configuration, this often leads to a reliance on lengthy terminal errors that can be expensive from a context-switching perspective.

{% video local /img/snippets/cypress-interactive-debugging.mp4 %}

### Faster feedback loops

> When it comes to your end-to-end tests, being able to see your tests as they run is critical to allowing you to iterate faster with confidence. Though this is not possible out of the box with Protractor, Cypress tests automatically re-run within the browser whenever they are saved which allows you to iterate faster with confidence.

Author: Great suggestion!

### Time travel through tests

> Rather than trying to decipher errors inside of your terminal, with Cypress you can debug your tests just like you would debug your application normally. In addition, Cypress allows you to see snapshots of your Angular application as the test is executed so you can debug your applications more easily.

Author: Thank you for the suggestion!

### Easily automate screenshots

> Another powerful feature of end-to-end testing is the ability to capture screenshots and recording. While possible in both tools, Protractor requires you to manually tap into Node.js and has limited configuration options.

Author: 👍

**Before: Protractor**

```javascript
const fs = require('fs')

describe('Dashboard', () => {
  it('should render dashboard', () => {
    browser.get('/dashboard')
    browser.takeScreenshot().then((png) => {
      const stream = fs.createWriteStream(filename)

      writeScreenShot(png, 'dashboard.png')
      stream.write(new Buffer(png, 'base64'))
      stream.end()
    })
  })
})
```

**After: Cypress**

```javascript
describe('Dashboard', () => {
  it('should render dashboard', () => {
	  cy.visit('/dashboard')
	  cy.screenshot('dashboard.png')
  })
})
```

> For more information on what configuration options are available, be sure to check out the [official Cypress documentation on the screenshot command](https://docs.cypress.io/api/commands/screenshot.html).

Author: Thanks for the suggestion!

## FAQs

### Do I have to replace all of my tests with Cypress immediately?

Absolutely not. While it might sound ideal to replace Protractor entirely, we know that most projects and teams still need to deliver features to customers and it's unrealistic to halt the team's progress.

### Can Protractor and Cypress coexist in the same app?

Yes! Your Protractor tests would continue to live in the `e2e` directory that Angular CLI scaffolded while all Cypress tests would live in a sibling folder named `cypress`.

```
.
├── cypress
├── e2e
├── src
├── .editorconfig
├── .gitignore
├── angular.json
├── browserslist
├── cypress.json
├── karma.conf.js
├── package.json
├── README.md
├── tsconfig.app.json
├── tsconfig.json
├── tsconfig.spec.json
└── tslint.json
```

In fact, as you work through migrating to Cypress, we believe that progressively enhancing your e2e tests with Cypress is the best path forward to ensure that feature development is not impacted.

> ## Getting Started

Reviewer: Make this another main heading

```suggestion
# Getting Started
```

Like any library you would add to your application, we start by adding it as a dependency to our `package.json`.

> ```
> npm install cypress -D
> ```

Reviewer: specify shell language for correct styling

```suggestion
```shell
npm install cypress -D
```
```

Then, since Cypress can run in parallel with your application, let's install Concurrently to simplify our npm script.

> ```
> npm install concurrently -D
> ```

Reviewer: specify shell language for correct styling.

```suggestion
```shell
npm install concurrently -D
```
```

Then we will update our `package.json` with the a script to run Cypress and your Angular app simultaneously:

```js
"cypress": "concurrently \"ng serve\" \"cypress open\""
```

```js
// Example config in package.json
{
  "scripts": {
    ...
    "e2e": "ng e2e",

>     "e2e:debug": "node --inspect-brk ./node_modules/.bin/protractor ./e2e/protractor.conf.js",

Reviewer: Beware of horizontal scrollbars if you can manage to avoid them. 

![Screen Shot 2020-05-21 at 3 55 57 PM](https://user-images.githubusercontent.com/1271364/82544768-a1705e00-9b7b-11ea-8c2c-14a5367e7f4b.png)

Reviewer 2: To make this line shorter we can use `$(npm bin)` over `./node_modules/.bin`.

Reviewer 2: @bencodezen does Protractor always need `portractor.conf.js` passed in explicitly or can it find it in the default location?

Author: @amirrustam Interestingly enough. It defaults to a protractor.conf.js file in the root directory, but the Angular CLI scaffolds that file in the e2e directory. 😆

>     "cypress": "concurrently \"ng serve\" \"cypress open\""

Author: ```suggestion
    "cypress:open": "concurrently \"ng serve\" \"cypress open\""
```

    "cypress": "concurrently \"ng serve\" \"cypress open\""
  },
  "dependencies": { ... },
  "devDependencies": { ... }
}
```

Now, when we run:

```bash
npm run cypress
```

It will start up Cypress and our Angular app at the same time!

{% video local /img/snippets/npm-run-cypress.mp4 %}

> ## Essentials

Reviewer: make this a main section heading too - need to update all subheadings to be h2 and h3 respectively (this will break things if you don't update subheadings! so make sure to update)
 
```suggestion
# Essentials
```

> npm run cypress

Author: ```suggestion
npm run cypress:open
```

**Before: Protractor**

```js
// Unable to get this to work just yet
describe('show stubbing example', () => {
  it('stubs a call to the login API', () => {
    browser.addMockModule('modName', function() {
      🔥 angular.module('modName', []).value('foo', 'bar');
    });		
  })
})
```

**After: Cypress**

```js
describe('show stubbing scaffold', () => {
  it('stubs a call to the login API', () => {
    cy.stub(api, 'login').resolves('user-id-123')
  })
})
```

With the `cy.stub()` method, you can easily replace a function, record its usage, and even control its behavior.

For more information, check out our [official documentation on stubbing](https://docs.cypress.io/api/commands/stub.html#Syntax).

### Navigating Websites

When you want to visit a page, you can do so with the following code:

**Before: Protractor**

```js
it('visits a page', () => {
  browser.get('/about')
  browser.navigate().forward()
  browser.navigate().back()
})
```

**After: Cypress**

```js
it('visit a non-Angular page', () => {

>   cy.visit('/about')

Author: ```suggestion
  cy.visit('http://localhost:4200/about')
```

exports.config = {

  seleniumAddress: 'https://cloudName.perfectomobile.com/nexperience/perfectomobile/wd/hub',

  // Capabilities to be passed to the webdriver instance.
  capabilities: {
    browserName: 'chrome',

    // allows different specs to run in parallel.
    // If this is set to be true, specs will be sharded by file
    // (i.e. all files to be run by this set of capabilities will run in parallel).
    // Default is false.
    sharedTestFiles: true,

    // Maximum number of browser instances that can run in parallel for this
    // set of capabilities. This is only needed if shardTestFiles is true.
    // Default is 1.
    maxInstances: 2,

    // Cloud capabilities
    user: 'user@perfectomobile.com',
    password: 'password',

    // Device capabilities
    platformName: 'Android',
    manufacturer: 'Samsung',
    model: 'Galaxy S5'
  },

  // Framework to use. Jasmine is recommended.
  framework: 'jasmine',

  // Spec patterns are relative to the current working directly when
  // protractor is called.
  specs: ['spec.js', 'spec2.js'],

  // Options to be passed to Jasmine.
  jasmineNodeOpts: {
    showColors: true,
    defaultTimeoutInterval: 30000
  }
}
```

> However, with Cypress, all you need to do is pass the `--parallel` and `--record` flag to `cypress run`, and it will take care of the rest for you:

Reviewer: "all you need to do" is kind of similar to using phrasing "just do this", which we do not allow in the docs because it implies something is very simple for everyone, so leads to extra frustration when the reader encounters issues implementing it.

```suggestion
However, with Cypress, you can pass the {% url "`--parallel`" command-line#cypress-run-parallel %} and {% url "`--record`" command-line#cypress-run-record-key-lt-record-key-gt %} flag to `cypress run`, and your tests will automatically run in parallel:
```

Author: Great point. Thanks for the suggestion!

**After: Cypress**

> ```bash
> $ cypress run --record --parallel

Reviewer: Do not include `$` within terminal text - we want the text to be fully copy/pastable. Adding the `shell` as the language will add a `$` to the styles later (which will not be copied if user copies) and have the correct styling.

```suggestion
```shell
cypress run --record --parallel
```

**Before**

![Screen Shot 2020-05-21 at 3 34 31 PM](https://user-images.githubusercontent.com/1271364/82542845-aa136500-9b78-11ea-9603-2c8f77b68563.png)

**After**

![Screen Shot 2020-05-21 at 3 35 02 PM](https://user-images.githubusercontent.com/1271364/82542857-af70af80-9b78-11ea-8c07-ac6b1df30346.png)

```

For more information, check out our [official docs on parallelization](https://docs.cypress.io/guides/guides/parallelization.html#Overview)!

### Element Explorer

> For those who are big fans of [Protractor's Element Explorer functionality](https://www.protractortest.org/#/debugging#enabled-control-flow), Cypress also provides you with a Selector Playground that allows you to:

Reviewer: ```suggestion
For those who are big fans of {% url "Protractor's Element Explorer functionality" https://www.protractortest.org/#/debugging#enabled-control-flow %}, Cypress also provides you with a {% url "Selector Playground" test-runner#Selector-Playground %} that allows you to:
```

- Determine a unique selector for an element
- See what elements match a given selector
- See what element matches a string of text

> For more information, check out our official docs on the [Selector Playground](https://docs.cypress.io/guides/core-concepts/test-runner.html#Selector-Playground).

Reviewer: url tag

> ## Next Steps

Reviewer: ```suggestion
# Next Steps
```

> For more information on how to create end-to-end tests with Cypress, be sure to check out [our official documentation here](https://docs.cypress.io/guides/overview/why-cypress.html).

Reviewer: url tag

{% note warning %}

> If you see any inaccuracies with this or feel something has been misrepresented, please [submit an issue here](https://github.com/cypress-io/cypress-documentation/issues/new).

Reviewer: url tag

> {% endnote %}

Reviewer: We typically have a See Also section at the bottom of guides - if you think there are some things to link to from here that may be helpful when they make it to the end. See other docs.

{% video local /img/snippets/cypress-interactive-debugging.mp4 %}

### Faster feedback loops

When it comes to your end-to-end tests, being able to see your tests as they run is critical to allowing you to iterate faster with confidence. Though this is not possible out of the box with Protractor, Cypress tests automatically re-run within the browser whenever they are saved which allows you to iterate faster with confidence.

### Time travel through tests

Rather than trying to decipher errors inside of your terminal, with Cypress you can debug your tests just like you would debug your application normally. In addition, Cypress allows you to see snapshots of your Angular application as the test is executed so you can debug your applications more easily.

### Easily automate screenshots

Another powerful feature of end-to-end testing is the ability to capture screenshots and recording. While possible in both tools, Protractor requires you to manually tap into Node.js and has limited configuration options.

**Before: Protractor**

```javascript
const fs = require('fs')

describe('Dashboard', () => {
  it('should render dashboard', () => {

>     i

Reviewer 3: @bencodezen Should this "i" be on this line?

Author: Not at all. Thanks for catching this!

	})
})
```

But Cypress has one additional feature that can make a critical difference in the reliability of your tests' assertions: [retry-ability](https://docs.cypress.io/guides/core-concepts/retry-ability.html). When your test fails an assertion or command, Cypress will mimic a real user with build-in wait times and multiple attempts at asserting your tests in order to minimize the amount of false negatives / positives. 

In the example above, if the submit link does not appear on the page at the exact moment when Protractor runs the test (which can be due to any number of factors including API calls, slow browser rendering, etc.), your test will fail. However, Cypress factors these conditions into its assertions and will only fail if the time goes beyond a reasonable amount.

You can learn more about how Cypress handles [assertions in our official documentation](https://docs.cypress.io/guides/references/assertions.html).

### Network Spies

By default, Protractor utilizes Jasmine's `spyOn` function to call a method and test that it was called.

```js
// paymentApi.js
export default {
  submit() { ... }
}
```

**Before: Protractor**

```js

> // Protractor

Reviewer 3: Is this comment redundant with header on line 345?

Author: Totally missed this. Thank you!

import paymentApi from './paymentApi.js'

describe('Protractor payment example', () => {
  it('verifies payment is only sent once', () => {
    spyOn(paymentApi, 'submit')
    browser.get('/payment')
    expect(paymentApi.submit).toHaveBeenCalledTimes(1)
  })
})
```

Fortunately for us, creating a spy in Cypress is almost identical!

**After: Cypress**

```js

> // Cypress

Reviewer 3: Is this redundant with header on line 362?

Author: Good catch!
